src/jaxsim/rbda/contacts/relaxed_rigid.py

```diff
@@ -0,0 +1,371 @@
+
+        return RelaxedRigidContactsParams.build(
+            time_constant=time_constant,
+            damping_coefficient=damping_coefficient,
+            d_min=d_min,
+            d_max=d_max,
+            width=width,
+            midpoint=midpoint,
+            power=power,
+            stiffness=stiffness,
+            damping=damping,
+            mu=mu,
+        )
+
+    def valid(self) -> bool:
+        return bool(
+            jnp.all(self.time_constant >= 0.0)
+            and jnp.all(self.damping_coefficient > 0.0)
+            and jnp.all(self.d_min >= 0.0)
+            and jnp.all(self.d_max <= 1.0)
+            and jnp.all(self.d_min <= self.d_max)
+            and jnp.all(self.width >= 0.0)
+            and jnp.all(self.midpoint >= 0.0)
+            and jnp.all(self.power >= 0.0)
+            and jnp.all(self.mu >= 0.0)
+        )
+
+
+@jax_dataclasses.pytree_dataclass
+class RelaxedRigidContactsState(ContactsState):
+    """Class storing the state of the relaxed rigid contacts model."""
+
+    def __eq__(self, other: RelaxedRigidContactsState) -> bool:
+        return hash(self) == hash(other)
+
+    @staticmethod
+    def build_from_jaxsim_model(
+        model: js.model.JaxSimModel | None = None,
+    ) -> RelaxedRigidContactsState:
+        """Build a `RelaxedRigidContactsState` instance from a `JaxSimModel`."""
+        return RelaxedRigidContactsState.build()
+
+    @staticmethod
+    def build() -> RelaxedRigidContactsState:
+        """Create a `RelaxedRigidContactsState` instance"""
+
+        return RelaxedRigidContactsState()
+
+    @staticmethod
+    def zero(model: js.model.JaxSimModel) -> RelaxedRigidContactsState:
+        """Build a zero `RelaxedRigidContactsState` instance from a `JaxSimModel`."""
+        return RelaxedRigidContactsState.build()
+
+    def valid(self, model: js.model.JaxSimModel) -> bool:
+        return True
+
+
+@jax_dataclasses.pytree_dataclass
+class RelaxedRigidContacts(ContactModel):
+    """Rigid contacts model."""
+
+    parameters: RelaxedRigidContactsParams = dataclasses.field(
+        default_factory=RelaxedRigidContactsParams
+    )
+
+    terrain: jax_dataclasses.Static[Terrain] = dataclasses.field(
+        default_factory=FlatTerrain
+    )
+
+    def compute_contact_forces(
+        self,
+        position: jtp.Vector,
+        velocity: jtp.Vector,
+        model: js.model.JaxSimModel,
+        data: js.data.JaxSimModelData,
+    ) -> tuple[jtp.Vector, tuple[Any, ...]]:
+
+        def _detect_contact(x: jtp.Array, y: jtp.Array, z: jtp.Array) -> jtp.Array:
+            x, y, z = jax.tree_map(jnp.squeeze, (x, y, z))
+
+            n̂ = self.terrain.get_normal_at(x=x, y=y).squeeze()
+            h = jnp.array([0, 0, z - model.terrain.get_height_at(x=x, y=y)])
+
+            return jnp.array([0, 0, jnp.dot(h, n̂)], dtype=float)
+
+        # Compute the activation state of the collidable points
+        position = jax.vmap(_detect_contact)(*position.T)
```

Comment:
Here you are overriding the `position` argument with a new variable that it seems storing the penetration depth $\delta$. Did I get it right? I'd probably change the variable name to be more clear and avoid making the mistake to use it later as ${}^W \mathbf{p}_C$.

Reply:
The original `position` contains the 3D positions of all the collidable points in the inertia-fixed reference frame. Here I'm computing the penetration with respect to the terrain that can be described by a generic heightmap. It could make sense to change the variable name, thank you!

Reply:
Done in https://github.com/ami-iit/jaxsim/pull/223/commits/30a6e4da8b61577cac23a787c1015804dc0089ee